### PR TITLE
[tagger/tagstore] Add DCA implementation that reduces mem usage

### PR DESCRIPTION
**Pull request by davidor**
### What does this PR do?

This PR reduces the memory used by the tagger when running in the Cluster Agent.

The tagger stores the tags by source and also stores the final result of aggregating them according to the priorities defined. This is a valid use case for the node agent because a single tagger entity can be reported by multiple sources. For example, a container might be reported by containerd but also by the Kubelet. Keeping the tags by source is used to merge them applying the priorities defined. Also `agent tagger-list` and some telemetry metrics show the tag source.

The case of the Cluster Agent is a bit different because tags only come from a single workloadmeta collector (kubeapiserver). That means that a single entity is not reported by multiple sources. That's why in this case we can reduce the memory used by not storing the tags by source.

This PR achieves that goal by defining a `EntityTags` interface that is responsible for storing the tags for a tagger entity and provides two implementations, one that handles the case of multiple sources (current implementation) and another one that assumes a single source (new optimization for the DCA).

This memory optimization is important in the Cluster Agent when the option to collect Kubernetes tags is enabled (`cluster_agent.collect_kubernetes_tags`) because in large clusters  the memory used to store the pod tags can be quite high.

### Describe how to test/QA your changes

For the DCA try with the option to collect Kubernetes tags enabled (`datadog.clusterTagger.collectKubernetesTags` option in the helm chart). Check in our internal clusters using profiles that the memory used by the tagger has decreased. Do also a quick verification of `agent tagger-list`.

For the Agent, a sanity check should be enough. Deploy the agent and check with `agent tagger-list` that tags are correct.

### CODE REVIEW

**Review by maycmlee — APPROVED** [review]
Just a small suggestion, but approving.

File: comp/core/tagger/taggerimpl/tagstore/entity_tags.go [hunk 1]
```diff
@@ -6,18 +6,55 @@
...
+	"github.com/DataDog/datadog-agent/pkg/util/flavor"
 )
 
-// EntityTags holds the tag information for a given entity. It is not
-// thread-safe, so should not be shared outside of the store. Usage inside the
-// store is safe since it relies on a global lock.
-type EntityTags struct {
+// This file defines the EntityTags interface which contains the tags for a
+// tagger entity.
+//
```
Comment (adel121) [inline comp/core/tagger/taggerimpl/tagstore/entity_tags.go:22]: should we still mention that it is not thread-safe and thus should not be shared outside the store? (Similar to the old comment)

File: comp/core/tagger/taggerimpl/tagstore/entity_tags.go [hunk 2]
```diff
@@ -147,3 +200,160 @@ func (e *EntityTags) shouldRemove() bool {
...
+
+func (e *EntityTagsWithMultipleSources) deleteSource(source string, expiryDate time.Time) {
+	tags, ok := e.sourceTags[source]
+	if !ok {
+		return
+	}
+
+	tags.expiryDate = expiryDate
+	e.sourceTags[source] = tags
+}
```
Comment (adel121) [inline comp/core/tagger/taggerimpl/tagstore/entity_tags.go:246]: Some questions about this method:
- Why is the name `deleteSource` while it doesn't actually delete the source, it only sets the expiration date for the source. Do you think `setSourceExpiration` is more relevant to what the method does?
- why do we need to run the last line of this method ( `e.sourceTags[source] = tags`)? It seems like we are setting `e.sourceTags[source]` to itself. Am I missing something?
-

File: comp/core/tagger/taggerimpl/tagstore/entity_tags.go [hunk 3]
```diff
@@ -75,7 +93,26 @@ func (e *EntityTags) toEntity() types.Entity {
 	}
 }
 
-func (e *EntityTags) computeCache() {
+func (e *EntityTagsWithMultipleSources) getStandard() []string {
+	tags := []string{}
+	for _, t := range e.sourceTags {
+		tags = append(tags, t.standardTags...)
```
Comment (adel121) [inline comp/core/tagger/taggerimpl/tagstore/entity_tags.go:99]: shouldn't we make sure to remove duplication here?

File: comp/core/tagger/taggerimpl/tagstore/entity_tags.go [hunk 4]
```diff
@@ -147,3 +200,160 @@ func (e *EntityTags) shouldRemove() bool {
...
+
+func (e *EntityTagsWithSingleSource) tagsForSource(source string) *sourceTags {
+	if source != e.source {
+		return nil
+	}
+
+	return &sourceTags{
+		lowCardTags:          e.cachedLow.Get(),
+		orchestratorCardTags: e.cachedAll.Slice(e.cachedLow.Len(), e.cachedOrchestrator.Len()).Get(),
+		highCardTags:         e.cachedAll.Slice(e.cachedOrchestrator.Len(), e.cachedAll.Len()).Get(),
```
Comment (adel121) [inline comp/core/tagger/taggerimpl/tagstore/entity_tags.go:309]: nit: I don't think it is necessary to specify the upper bound of the Slice here.

File: comp/core/tagger/taggerimpl/tagstore/entity_tags.go [hunk 5]
```diff
@@ -147,3 +200,160 @@ func (e *EntityTags) shouldRemove() bool {
...
+	}
+}
+
+func (e *EntityTagsWithSingleSource) tagsBySource() map[string][]string {
+	return map[string][]string{e.source: e.cachedAll.Get()}
+}
+
+func (e *EntityTagsWithSingleSource) setTagsForSource(source string, tags sourceTags) {
+	if source != e.source {
+		return
```
Comment (adel121) [inline comp/core/tagger/taggerimpl/tagstore/entity_tags.go:321]: should we emit some log (or even a panic) here to indicate that something doesn't look right.

File: comp/core/tagger/taggerimpl/tagstore/entity_tags.go [hunk 6]
```diff
@@ -147,3 +200,160 @@ func (e *EntityTags) shouldRemove() bool {
...
+	e.cachedAll = cached
+	e.cachedLow = cached.Slice(0, len(tags.lowCardTags))
+	e.cachedOrchestrator = cached.Slice(0, len(tags.lowCardTags)+len(tags.orchestratorCardTags))
+}
+
+func (e *EntityTagsWithSingleSource) sources() []string {
+	return []string{e.source}
+}
+
+func (e *EntityTagsWithSingleSource) deleteSource(source string, expiryDate time.Time) {
```
Comment (adel121) [inline comp/core/tagger/taggerimpl/tagstore/entity_tags.go:342]: same question as before (for `EntityTagsWithMultipleSource`)

**Review by adel121 — COMMENTED** [review]
Thanks @davidor, great PR!

Left some comments and questions

File: comp/core/tagger/taggerimpl/tagstore/entity_tags.go (hunk 4 above)
Comment (davidor) [reply comp/core/tagger/taggerimpl/tagstore/entity_tags.go:311]: It's calling this `.Slice` function: https://github.com/DataDog/datadog-agent/blob/0634e14562761cee584c7ad81c794b473908c0b2/pkg/tagset/hashed_tags.go#L27
It seems to be necessary.

File: comp/core/tagger/taggerimpl/tagstore/entity_tags.go (hunk 3 above)
Comment (davidor) [reply comp/core/tagger/taggerimpl/tagstore/entity_tags.go:100]: I didn't change this part. It's like this in the main branch:
https://github.com/DataDog/datadog-agent/blob/0634e14562761cee584c7ad81c794b473908c0b2/comp/core/tagger/taggerimpl/tagstore/entity_tags.go#L37

File: comp/core/tagger/taggerimpl/tagstore/entity_tags.go (hunk 5 above)
Comment (davidor) [reply comp/core/tagger/taggerimpl/tagstore/entity_tags.go:324]: I added this in the new commit 👍

File: comp/core/tagger/taggerimpl/tagstore/entity_tags.go (hunk 2 above)
Comment (davidor) [reply comp/core/tagger/taggerimpl/tagstore/entity_tags.go]: I renamed the function in the new commit :+1:

File: comp/core/tagger/taggerimpl/tagstore/entity_tags.go (hunk 6 above)
Comment (davidor) [reply comp/core/tagger/taggerimpl/tagstore/entity_tags.go]: Renamed in the new commit :+1:

**Comment by adel121** [comment]
LGTM 🙇

**Review by adel121 — APPROVED** [review]

**Pull request merged**